### PR TITLE
DAOS-9217 daos: add rf pda pool and container properties

### PR DESCRIPTION
**Pull request by wangshilong**
Similarly to the EC cell size, we should have a default pool property for the
redundancy factor (RF) which is inherited to all containers (if not specify at
container creation time) in the pool. Default pool rf is 0.

Introduce another two new properties for pool and container as the user interfaces:

ec_pda=<integer>: Performance Domain Affinity Level of EC object.
rp_pda=<integer>: Performance Domain Affinity Level of replicated object.

The user can specify the PDA by setting these properties for either the pool or
the container or both: if the user specifies these properties only for a pool,
then all containers within this pool will inherit them from the pool; if the
user specifies them for both pool and container, then container properties will
the pool properties.

Default value of rp_pda is 3, the placement algorithm should place two-way and
three-way replication object within a PD; for those object classes with more
than 3 replicas, DAOS will place three replicas within a PD and switch to another PD.

Default value of EC_PDA is 1, the placement algorithm always tries to scatter
shards of EC object to different PDs.

Signed-off-by: Wang Shilong <shilong.wang@intel.com>

### CODE REVIEW

**Comment by wangshilong** [comment]
ping reviewers

**Review by mjmac — COMMENTED** [review]
Looks good overall to me. Minor nit in the Go property handler, wouldn't block on it, but there are also some documentation warnings in daos_prop.h that should probably be fixed.

**Comment by mjmac** [comment]
Added Ding and Makito for review of the test code.

File: src/tests/ftest/pool/rf.py
Comment (shimizukko) [inline src/tests/ftest/pool/rf.py:24]:
```suggestion
            expected_value (int): expected container rf value
```
typo

**Review by shimizukko — DISMISSED** [review]
I think you should run pool/rf.py in CI. You can add the following tag in the commit message:
Test-tag-hw-large: pr pool_rf_property

If you want to run just this test and not other PR tests,
Test-tag: pool_rf_property

**Review by shimizukko — APPROVED** [review]

**Review by shimizukko — APPROVED** [review]

**Review by mjmac — COMMENTED** [review]
Code LGTM, but there is an NLT error.

**Review by shimizukko — APPROVED** [review]

**Review by mjmac — DISMISSED** [review]
LGTM, but I don't know what the ramifications of bumping the pool version are. If @johannlombardi and/or @gnailzenh are OK with it, then so am I.

File: src/pool/srv_layout.h
Comment (kccain) [inline src/pool/srv_layout.h:29]: just a quick comment (I haven't reviewed this patch, for example haven't looked at it from a layout version compatibility perspective).

In this patch the DAOS_CONT_MD_VERSION will need to take the same value as DAOS_POOL_MD_VERSION due to https://daosio.atlassian.net/browse/DAOS-8918. So they must be changed together for now.

Comment (wangshilong) [reply src/pool/srv_layout.h:29]: I see, let me bump version together now then.

**Review by shimizukko — APPROVED** [review]

**Review by shimizukko — APPROVED** [review]

**Comment by wangshilong** [comment]
> Did this get merged with #7878? I think we would need to merge #7878 first and then rebase this PR.

Yup, I merge them into one PR, do you think it better to split them?

**Comment by daltonbohning** [comment]
> > Did this get merged with #7878? I think we would need to merge #7878 first and then rebase this PR.
> 
> Yup, I merge them into one PR, do you think it better to split them?

I think it's okay if necessary, but merging means re-reviewing, since the commits show up as "new" commits on this PR.

**Comment by wangshilong** [comment]
> > > Did this get merged with #7878? I think we would need to merge #7878 first and then rebase this PR.
> > 
> > 
> > Yup, I merge them into one PR, do you think it better to split them?
> 
> I think it's okay if necessary, but merging means re-reviewing, since the commits show up as "new" commits on this PR.

I see..I can restore #7878

**Comment by wangshilong** [comment]
> > > > Did this get merged with #7878? I think we would need to merge #7878 first and then rebase this PR.
> > > 
> > > 
> > > Yup, I merge them into one PR, do you think it better to split them?
> > 
> > 
> > I think it's okay if necessary, but merging means re-reviewing, since the commits show up as "new" commits on this PR.
> 
> I see..I can restore #7878

It shows merged..any way i could reopen it?

**Comment by daltonbohning** [comment]
> > > > > Did this get merged with #7878? I think we would need to merge #7878 first and then rebase this PR.
> > > > 
> > > > 
> > > > Yup, I merge them into one PR, do you think it better to split them?
> > > 
> > > 
> > > I think it's okay if necessary, but merging means re-reviewing, since the commits show up as "new" commits on this PR.
> > 
> > 
> > I see..I can restore #7878
> 
> It shows merged..any way i could reopen it?

Ah, #7878 was a PR into THIS PR, not master. Ideally, I think we should have either (1) merged #7878 into master as an individual commit, (2) merged the two patches in the beginning, or (3) used a feature branch.
Someone might have a different opinion, but I think the simplest path forward is to re-review here, and we should make it clear to the gatekeeper that the final squashed commit needs the message for both DAOS-9217 and DAOS-9550.

**Comment by johannlombardi** [comment]
Please merge master.

**Review by shimizukko — DISMISSED** [review]

File: src/container/srv_layout.h
Comment (liuxuezhao) [inline src/container/srv_layout.h:96]: should be uint64?

File: src/client/api/container.c
Comment (wangshilong) [reply src/client/api/container.c:100]: See above if redunc factor specified, no need inherit from pool，we won't come here, and there is rf.py test case covered this.

Comment (liuxuezhao) [reply src/client/api/container.c:100]: ok, i see, thx

File: src/pool/cli.c
Comment (wangshilong) [reply src/pool/cli.c:2797]: Originally i hit problem calling sync in async, so i moved this out from task, see daos_cont_inherit_redunc_fac() is called in daos_cont_create2() that happens before dc_task_create(), so it probably ok?.

Extend pool_connect_out to return more pool information could break compatibity between older clients and new server, i don't have strong objection for that though.

Comment (liuxuezhao) [reply src/pool/cli.c:2797]: I think that generally make sync call with RPC within async API is not good, and it could affect cont_create's performance.
But probably fine to leave it to be resolved in following other patch, I don't want to block this patch's land.

**Review by liuxuezhao — APPROVED** [review]

File: src/pool/cli.c
Comment (wangshilong) [reply src/pool/cli.c:2797]: make sense, thanks!

File: src/include/daos_prop.h
Comment (jolivier23) [inline src/include/daos_prop.h:87]: should these be at the end?  You are changing the value of DAOS_PROP_PO_POLICY.  Otherwise, this is actually an ABI breaking change and a major version update, no?

File: src/include/daos_prop.h
Comment (jolivier23) [inline src/include/daos_prop.h:77]: This should be a minor version update for the API

**Review by jolivier23 — CHANGES_REQUESTED** [review]

File: src/include/daos_prop.h
Comment (wangshilong) [reply src/include/daos_prop.h:87]: This is a mischanged by rebasing, let me fix it.

**Review by shimizukko — APPROVED** [review]
Please add test description here.
https://daosio.atlassian.net/wiki/spaces/DAOS/pages/2194899612/Pool+Test+Plan

**Review by liuxuezhao — DISMISSED** [review]

**Comment by liuxuezhao** [comment]
@wangshilong  just though one point that for the RF, and PDA property should only be set when pool create or container create right? so, in the code should forbidden in xxx_set_prop() path, need some extra code for that checking.

**Comment by liuxuezhao** [comment]
> @wangshilong just though one point that for the RF, and PDA property should only be set when pool create or container create right? so, in the code should forbidden in xxx_set_prop() path, need some extra code for that checking.

can refer the code in dc_cont_set_prop() the check for DAOS_PROP_CO_EC_CELL_SZ, just FYI.

**Comment by wangshilong** [comment]
> > @wangshilong just though one point that for the RF, and PDA property should only be set when pool create or container create right? so, in the code should forbidden in xxx_set_prop() path, need some extra code for that checking.

Yup, this is missed.

> 
> can refer the code in dc_cont_set_prop() the check for DAOS_PROP_CO_EC_CELL_SZ, just FYI.
Sure, let me check and possibly add a test as well.

File: src/pool/cli.c
Comment (liuxuezhao) [reply src/pool/cli.c:2797]: np, to avoid RPC protocol change, probably can do another pool query inside client's pool connect, so can cache that prop and need not query again later. just FYI.

Comment (wangshilong) [reply src/pool/cli.c:2822]: I pushed a simple commit to cache rf prop, but i found this caching only work for api which pool handle can be shared inside same process etc, i am wondering to make pool connect more extendable, it might make sense to introduce a POOL_CONNECT_V2.

**Review by liuxuezhao — APPROVED** [review]

File: src/pool/srv_pool.c
Comment (wangshilong) [reply src/pool/srv_pool.c:3639]: PDA is only pool property, not container, maybe need check rf you meaned?

**Review by shimizukko — APPROVED** [review]

File: src/pool/srv_pool.c
Comment (liuxuezhao) [reply src/pool/srv_pool.c:3639]: there is with "DAOS_PROP_CO_EC_PDA" and "DAOS_PROP_CO_RP_PDA"? with risk of changed by daos_cont_set_prop() after container created/opened?

Comment (wangshilong) [reply src/pool/srv_pool.c:3639]: you are right, my mind was broken, should have same check for it.

Comment (liuxuezhao) [reply src/pool/srv_pool.c:3639]: np, can add a check at following patch, need not refresh again. thx

**Comment by liuxuezhao** [comment]
@jolivier23 could you please check if the "request change" still necessary? thx

File: src/control/cmd/dmg/pool.go
Comment (mchaarawi) [inline src/control/cmd/dmg/pool.go:570]: don't use pda for messages. user doesn't know all of our acronymps.

File: src/include/daos_prop.h
Comment (mchaarawi) [inline src/include/daos_prop.h:83]: no need to split comments on 2 lines. our max line width is 100 characters :-)

**Review by shimizukko — APPROVED** [review]

File: src/include/daos_prop.h
Comment (kjacque) [inline src/include/daos_prop.h:130]: Looks like each of these values is defined twice. Is it a copy/paste error or did you intend one of the pairs to be DAOS_PROP_CO_... instead?

File: src/control/drpc/property.go
Comment (kjacque) [inline src/control/drpc/property.go:44]: Complete this comment please

File: src/control/drpc/property.go
Comment (kjacque) [inline src/control/drpc/property.go:48]: Not sure if gofmt will yell at you, but a space is usually expected after the "//" for comments.

File: src/control/drpc/property.go
Comment (kjacque) [inline src/control/drpc/property.go:104]: Since these are exported methods in Go (start with a capital letter), you should add a comment for each starting with the name of the function, similar to what was done for the LabelIsValid function above.

File: src/control/cmd/dmg/pool.go
Comment (kjacque) [inline src/control/cmd/dmg/pool.go:576]: OK for now, but I'm wondering if this should be defined in the pool property handler map instead, in the long term. Container props have a read-only flag, we could do something similar for pool props.

An aside, the message is a little awkward in English. Instead of "if pool is created," how about "on existing pool"?

**Review by kjacque — COMMENTED** [review]
I don't think any of my comments are blocking, but please address them in a follow-on PR.

File: src/include/daos_prop.h
Comment (wangshilong) [reply src/include/daos_prop.h]: removed duplicate

File: src/control/cmd/dmg/pool.go
Comment (wangshilong) [reply src/control/cmd/dmg/pool.go]: Sure, fixed comments, let's address Readonly in seperate PR.

**Review by shimizukko — DISMISSED** [review]

**Review by kjacque — DISMISSED** [review]

**Review by shimizukko — APPROVED** [review]

**Comment by wangshilong** [comment]
@daos-stack/daos-gatekeeper would you consider landing the PR? failure is known as DAOS-9860 thanks!

**Review by shimizukko — APPROVED** [review]

**Review by johannlombardi — APPROVED** [review]

**Pull request merged**
